### PR TITLE
[WIP] Cylindrical geofence

### PR DESCRIPTION
**Pull request by nmaggioni**
This PR enables the definition of a cylindrical geofence with both horizontal (distance from home) and vertical (altitude from home) boundaries, partially implementing #62.

New settings:
+ `geofence_radius`
+ `geofence_height`
+ `geofence_buffer`

Behavior:
1. Aircraft exceeds horizontal (`geofence_radius`, _red line below)_, vertical (`geofence_height`) or both boundaries.
2. RTH is engaged.
    - Pilot lets the aircraft fly home.
      + Usual RTH behavior.
    - Pilot moves the sticks and disengages RTH, thus proceeding outside the geofence.
      + Geofence gets disabled until the aircraft comes back inside the boundaries plus `geofence_buffer` meters _(blue line below is the buffer size, green the resulting boundary)_. This gives some more room to disengage RTH and get back on track, especially with MRs.

geofence

---

+ [ ] Field tested
+ [X] ~~Configurator PR~~ Should this feature be configurable via CLI only?

### CODE REVIEW

File: src/main/navigation/navigation.c [hunk 1]
```diff
@@ -3087,6 +3090,35 @@ void updateFlightBehaviorModifiers(void)
...
+
+void processGeofenceStatus(void)
+{
+    static bool geofenceTriggered = false;
+
+    if (!geofenceTriggered && isOutsideGeofence()) {
+        geofenceTriggered = true;
+        // No need to update status if F/S has already been enforced
+        if (getStateOfForcedRTH() == RTH_IDLE) {
+                activateForcedRTH();
```
Comment (nmaggioni) [inline src/main/navigation/navigation.c:3114]: Which is the right way to trigger RTH and/or check if it's already active? `activateForcedRTH()`, `failsafeActivate(FAILSAFE_RETURN_TO_HOME)`, `FLIGHT_MODE(FAILSAFE_MODE)`, or a combination of them? The functionality of `activateForcedRTH()` looks controversial to me.

**Comment by junwoo091400** [comment]
Nice picture man! But I do not understand this part:(about the Buffer)

"This gives some more room to disengage RTH and get back on track, especially with MRs."

Hmm can you explain real use case? I'm not sure if I understood it right, because currently i don't think that buffer is necessary.. Will look into the code..

**Comment by nmaggioni** [comment]
@junwoo091400 I didn't find a better way to explain it, but the buffer is needed to avoid hysteresis on the fringe of the geofence. [This](https://github.com/iNavFlight/inav/pull/4798/files#diff-b8166c625b87a27809555a2f13ba9c6bR3117) part of the code handles that feature, and here are two practical examples.

### Without buffer:
+ `geofence_radius` is set to 200m.
+ You fly out to 200m, RTH triggers.
+ As your craft is flying back, you move the sticks and disengage RTH. But your craft is now already less than 200m from home, so it's back inside the geofence.
+ You head back out to 200m, the cycle repeats because you keep bumping into RTH unless you "power through" the geofence boundary with enough speed to gain some extra distance before RTH engages and you manage to get out of it while still being >200m from home.

### With buffer:
+ `geofence_radius` is set to 200m, `geofence_buffer` to 50m.
+ You fly out to 200m, RTH triggers.
+ As your craft is flying back, you move the sticks and disengage RTH. If your craft is at least at `geofence_radius - geofence_buffer` meters from home (and thus 150m with this example config), it will still be considered outside the geofence. _So, in this case, you'll have 50m to safely disengage RTH._
+ You head back out to 200m and this time RTH won't engage, letting you go outside of the geofence.
+ As soon as you get back in a `geofence_radius - geofence_buffer` meters radius from home, the geofence is set back in place and the cycle repeats.

**Comment by junwoo091400** [comment]
> * You head back out to 200m, the cycle repeats because you keep bumping into RTH unless you "power through" the geofence boundary with enough speed to gain some extra distance before RTH engages and you manage to get out of it while still being >200m from home.

Ah, that makes a lot of sense. I see what you mean :) Thanks for the detailed explanation!

**Comment by nszceta** [comment]
Perhaps I am misunderstanding your logic but I think a much simpler and less error prone geofence algorithm would force RTH against the user's wishes until the machine is well within the geofence boundary. Only then is the ability to manually disable RTH allowed. The geofence itself should be toggled by a switch to take complete control of the machine as a last resort. Especially if the geofence is misconfigured such as if the diameter is impossibly small.

**Comment by nmaggioni** [comment]
@nszceta User misconfiguration can never be completely avoided, it's up to the pilot to set coherent parameters - **just like any other feature.** In my opinion _forcing_ RTH to be enabled for any any amount of time while the RX is still active is too dangerous, human judgment is fundamental in the use of this feature and placing it under a switch is not needed thanks to the buffer logic. Also do remember that RTH (F/S, in this case) will stay active until the pilot moves the sticks, so your suggestion pretty much matches the already implemented behavior if the pilot does nothing.

**Comment by moelski** [comment]
Hi !
Any chance that this will be included in INAV 2.3? 
This is an awesome feature for RC Training. No more fly away planes from the students :-)

**Comment by moelski** [comment]
Hmm no news on this topic?

**Comment by nAjss** [comment]
When to expect Geo-fencing available?

**Comment by nmaggioni** [comment]
I still need to finalize and test some code, most of the work is done but external situations are getting in my way.

**Comment by datamann** [comment]
Just let me know if you want me to test the feature

**Comment by mistyk** [comment]
Could you explain why you want to restrict your flight area ?

**Comment by nmaggioni** [comment]
@mistyk I think you should have read the other comments on both this PR and the related issue before asking that question.

**Comment by nmaggioni** [comment]
Ping, [see above](https://github.com/iNavFlight/inav/pull/4798#issuecomment-526943849).

**Comment by nmaggioni** [comment]
@Mateyhv A single buffer dimension is enough in my opinion, you'll still need some vertical space to maneuver to disable the RTH and obstacles might be in your way, so extra vertical space is crucial to avoid unwanted control interferences. Regarding the behavior: since at the geofence's boundary RTH will be enabled, it will behave just like you turned it on yourself.

Anyway, I encourage everyone to wait for an initial cut of this feature to be finalized, tested and merged before suggesting improvements.

**Comment by rsilk1949** [comment]
Fyi, Cylindrical geofence implementation in Arducopter:
https://ardupilot.org/copter/docs/common-ac2_simple_geofence.html

**Comment by b14ckyy** [comment]
As this is unfortunately stale since 4 years and we have a full 3D Vector based geozone feature with auto-navigation ready to be implemented, I will close this here to clean up the repo a bit.

**Pull request closed without merging**